### PR TITLE
change jaeger options to functional style

### LINKED ISSUE #146: Jaeger exporter should use functional options
Functional options would be more consistent with the rest of the code base.

Additionally, it would let us make certain configuration more clear, e.g., enabling a single `endpoint` option that would let you use _either_ a collector endpoint _or_ an agent endpoint.

### PR DESCRIPTION
**Pull request by paivagustavo**
Fixes #146 

With this i have separated the Options struct in 3:
1. General Options that apply to any jaeger exporter;
2. Collector Endpoint Options;
3. Agent Endpoint Options

Every exporter *must* pass one, and only one, endpoint (through the `EndpointOption`). The new signature is:
```
func NewExporter(endpointOption EndpointOption, opts ...Option) (*Exporter, error) {
```

Usage examples:
```go
	exporter, err := jaeger.NewExporter(
		jaeger.WithCollectorEndpoint("http://localhost:14268/api/traces"),
		jaeger.WithProcess(jaeger.Process{
			ServiceName: "trace-demo",
		}),
	)

	exporter, err = jaeger.NewExporter(
		jaeger.WithCollectorEndpoint(
			"http://localhost:14268/api/traces",
			jaeger.WithUsername("user"), // this only exists inside "WithCollectorEndpoint" option
			jaeger.WithPassword("password")), // this only exists inside "WithCollectorEndpoint" option
		jaeger.WithOnError(func(err error) {
			log.Fatal(err.Error())
		}),
	)

	exporter, err = jaeger.NewExporter(
		jaeger.WithAgentEndpoint("localhost:6831"),
		jaeger.WithBufferMaxCount(100),
	)
```

I have also extracted the upload to an interface:
```go
// batchUploader send a batch of spans to Jaeger
type batchUploader interface {
	upload(batch *gen.Batch) error
}
```

The `EndpointOption` generates an implementation of `batchUploader` based on the option passed by the user.

### CODE REVIEW

File: exporter/trace/jaeger/uploader.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	return func(o *CollectorEndpointOptions) {
+		o.password = password
+	}
+}
+
+// agentUploader implements batchUploader interface sending batches to
+// Jaeger through the UDP agent.
+type agentUploader struct {
+	client *agentClientUDP
+}
```
Comment (rghetia) [inline exporter/trace/jaeger/uploader.go]: add interface validation
```go
var _ batchUploader = (*agentUploader)(nil)
```

File: exporter/trace/jaeger/uploader.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	return a.client.EmitBatch(batch)
+}
+
+// collectorUploader implements batchUploader interface sending batches to
+// Jaeger through the collector http endpoint.
+type collectorUploader struct {
+	endpoint string
+	username string
+	password string
+}
```
Comment (rghetia) [inline exporter/trace/jaeger/uploader.go]: add interface validator.

File: exporter/trace/jaeger/jaeger.go [hunk 3]
```diff
@@ -34,55 +27,58 @@ import (
...
-
 	// Process contains the information about the exporting process.
 	Process Process
 
 	//BufferMaxCount defines the total number of traces that can be buffered in memory
 	BufferMaxCount int
 }
 
+// WithOnError sets the hook to be called when there is
+// an error occurred when uploading the stats data.
```
Comment (rghetia) [inline exporter/trace/jaeger/jaeger.go]: nit:s/stats/span/

**Comment by paivagustavo** [comment]
@rghetia I've changed what you request, could you review it again?

**Pull request merged**